### PR TITLE
add NetworkStatus in NetworkPlugin interface for kubelet to consume

### PR DESCRIPTION
**Pull request by freehan**
ref: #24692

cc: @thockin @dcbw

### CODE REVIEW

File: pkg/kubelet/kubelet.go [hunk 1]
```diff
@@ -2788,29 +2798,27 @@ func (kl *Kubelet) recordNodeStatusEvent(eventtype, event string) {
...
-				glog.Infof("Flannel server handshake failed %v", err)
-				return
-			}
-			kl.updatePodCIDR(podCIDR)
-		}
-		if err := ensureIPTablesMasqRule(kl.nonMasqueradeCIDR); err != nil {
-			err = fmt.Errorf("Error on adding ip table rules: %v", err)
-			glog.Error(err)
-			kl.runtimeState.setNetworkState(err)
+	if kl.flannelExperimentalOverlay {
```
Comment (yujuhong) [inline pkg/kubelet/kubelet.go]: Could we consolidate all the network state initialization/syncing to this function? E.g., if pod CIDR is required for the network plugins, we can still check whether pod CIDR is set and then call `setNetworkState()`. What do you think?

File: pkg/kubelet/kubelet.go [hunk 2]
```diff
@@ -3597,6 +3605,11 @@ func (kl *Kubelet) updatePodCIDR(cidr string) {
 		details := make(map[string]interface{})
 		details[network.NET_PLUGIN_EVENT_POD_CIDR_CHANGE_DETAIL_CIDR] = cidr
 		kl.networkPlugin.Event(network.NET_PLUGIN_EVENT_POD_CIDR_CHANGE, details)
+
+		// Assuming PodCIDR update is the only blocker for network plugin to bootstrap network.
+		// TODO: implement a better way to let network plugins report networkError back to kubelet
+		// Also consider seperate the runtimeState into Container Runtime state and Network Runtime state
+		kl.runtimeState.networkError = nil
```
Comment (yujuhong) [inline pkg/kubelet/kubelet.go]: You should call `kl.runtime.State.setNetworkState()`

Comment (freehan) [inline pkg/kubelet/kubelet.go]: urrrr. my bad.

File: pkg/kubelet/kubelet.go (hunk 1 above)
Comment (freehan) [inline pkg/kubelet/kubelet.go]: let me try

File: pkg/kubelet/runtime.go [hunk 3]
```diff
@@ -89,12 +89,8 @@ func (s *runtimeState) errors() []string {
 
 func newRuntimeState(
 	runtimeSyncThreshold time.Duration,
-	configureNetwork bool,
 ) *runtimeState {
-	var networkError error = nil
-	if configureNetwork {
-		networkError = fmt.Errorf("network state unknown")
-	}
+	var networkError = fmt.Errorf("network state unknown")
```
Comment (yujuhong) [inline pkg/kubelet/runtime.go]: nit: just initialize in line 97.

**Comment by yujuhong** [comment]
Looks good overall with a nit. Also, the kubelet unit test failed.

**Comment by freehan** [comment]
@k8s-bot e2e test this issue: #24642

**Comment by freehan** [comment]
Not sure why this happens
```
!!! 'gofmt -s' needs to be run on the following files: 
./pkg/api/v1/generated.pb.go
FAILED   ./hack/../hack/verify-gofmt.sh	7s
```

@k8s-bot test this issue: #IGNORE

**Comment by freehan** [comment]
@k8s-bot test this issue: #IGNORE

**Comment by dcbw** [comment]
What's the forward plan here for NetworkStatus()?  Is it intended to report plugin status out-of-band from setup/teardown operations?

**Comment by freehan** [comment]
@dcbw  Yes. Kubelet tries to keep track of network status and it already runs a `SyncNetworkStatus` loop. But network plugin currently does not has an interface to expose its own status. 

BTW, I want to rename the current `Status` interface to something like `GetPodNetworkStatus`

**Comment by yujuhong** [comment]
> BTW, I want to rename the current Status interface to something like GetPodNetworkStatus

+1 for renaming the method

**Comment by freehan** [comment]
PTAL

File: pkg/kubelet/network/plugins.go [hunk 4]
```diff
@@ -73,7 +73,10 @@ type NetworkPlugin interface {
 	TearDownPod(namespace string, name string, podInfraContainerID kubecontainer.ContainerID) error
 
 	// Status is the method called to obtain the ipv4 or ipv6 addresses of the container
-	Status(namespace string, name string, podInfraContainerID kubecontainer.ContainerID) (*PodNetworkStatus, error)
+	GetPodNetworkStatus(namespace string, name string, podInfraContainerID kubecontainer.ContainerID) (*PodNetworkStatus, error)
+
+	// NetworkStatus returns error if the network plugin is in error state
+	NetworkStatus() error
```
Comment (yujuhong) [inline pkg/kubelet/network/plugins.go]: nit: I think this can be rename to `Status()`

**Comment by freehan** [comment]
Renamed. PTAL

**Comment by yujuhong** [comment]
LGTM

**Comment by freehan** [comment]
Thanks!

**Comment by freehan** [comment]
@k8s-bot test this issue: #IGNORE

**Comment by freehan** [comment]
@k8s-bot test this issue: #24453

**Comment by dcbw** [comment]
@freehan I'm hoping that syncNetworkStatus() will eventually go away when kubenet replaces the configureCBR0 code.  But I guess the current use of the network plugin status for kubenet is OK, so LGTM too.

**Comment by freehan** [comment]
@k8s-bot test this issue: #24937

**Comment by freehan** [comment]
@k8s-bot test this issue: #IGNORE

**Comment by freehan** [comment]
This PR fixes cluster bootstrap problem in kubenet e2e suite. 
cc build cop: @dchen1107

**Pull request merged**